### PR TITLE
[mlir][Linalg] Add speculation for LinalgStructuredOps

### PR DESCRIPTION
**Pull request by Groverkss**
This patch adds speculation behavior for linalg structured ops, allowing them to be hoisted out of loops using LICM.

### CODE REVIEW

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp [hunk 1]
```diff
@@ -1202,6 +1203,20 @@ void GenericOp::getEffects(
   getGenericEffectsImpl(effects, cast<LinalgOp>(getOperation()));
 }
 
+static Speculation::Speculatability
+getGenericSpeculatabilityImpl(LinalgOp linalgOp) {
+  // Operands with value semantics are speculatable, while operands with memory
+  // semantics are not.
+  if (!linalgOp.hasPureTensorSemantics())
+    return Speculation::NotSpeculatable;
+  // The body of the op can still have speculation in it's region.
```
Comment (kuhar) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:1212]:
```suggestion
  // The body of the op can still have speculation in its region.
```

File: mlir/test/Transforms/loop-invariant-code-motion.mlir [hunk 2]
```diff
@@ -1118,3 +1118,48 @@ func.func @hoist_from_scf_while(%arg0: i32, %arg1: i32) -> i32 {
...
+    } -> tensor<128x128xf32>
+
+    %newacc = tensor.insert_slice %compute into 
+                                  %output[%i, 0][128, 128][1, 1] 
+                                  : tensor<128x128xf32> into tensor<?x128xf32>
+    scf.yield %newacc : tensor<?x128xf32>
+  }
+
+  func.return %final : tensor<?x128xf32>
+}
```
Comment (kuhar) [inline mlir/test/Transforms/loop-invariant-code-motion.mlir:1165]: Could we have a test with something to exercise the recursive aspect of speculablity? Maybe potential division by zero in the body?

Comment (Groverkss) [reply mlir/test/Transforms/loop-invariant-code-motion.mlir:1165]: added

**Review by kuhar — APPROVED** [review]
LGTM but let's wait for an approval from someone from the linalg side

**Review by dcaballe — COMMENTED** [review]
This is cool :)

I was wondering if, in addition to the tensor vs memref semantics, we should take into account the speculatability of the nested operations?

**Comment by Groverkss** [comment]
> This is cool :)
> 
> I was wondering if, in addition to the tensor vs memref semantics, we should take into account the speculatability of the nested operations?

I think we are accounting that. On tensor semantics, the speculatability of linalg ops is defined as recursively speculatable. I checked licm implementation, and it recursively checks every region if an op is recursively speculatable, so this should work. There is an example test added which does not hoist because a nested op could cause division by zero.

**Review by dcaballe — APPROVED** [review]
Awesome!

>  division by zero.

That's exactly the case I was thinking about. Sorry, I missed the CHECKs.

**Review by MaheshRavishankar — APPROVED** [review]
I don't know much about this but looks mostly right to me.

**Review by rengolin — APPROVED** [review]

**Pull request merged**